datasets/doc2dial/doc2dial.py

```diff
@@ -0,0 +1,167 @@
+                                {
+                                    "keys": datasets.Value("string"),   
+                                    "values": datasets.Value("string"), 
+                                }
+ 
+                            ),
+
+                            "utternace": datasets.Value("string"),
+                        }
+                    ),
+                }
+            ),
+            # No default supervised_keys (as we have to pass both question
+            # and context as input).
+            supervised_keys=None,
+            homepage=_HOMEPAGE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        
+        my_urls = _URLs[self.config.name]
+        print('my_url:',my_urls)
+        data_dir = dl_manager.download_and_extract(my_urls)
+
+        return [
+            datasets.SplitGenerator(name=datasets.Split.TRAIN, gen_kwargs={"filepath": os.path.join(data_dir, "doc2dial/v0.9/data/woOOD/doc2dial_dial_train.json")}),
+            datasets.SplitGenerator(name=datasets.Split.VALIDATION, gen_kwargs={"filepath": os.path.join(data_dir, "doc2dial/v0.9/data/woOOD/doc2dial_dial_dev.json")}),
+        ]
+
+    def _generate_examples(self, filepath):
+        """This function returns the examples in the raw (text) form."""
+
+        logging.info("generating examples from = %s", filepath)
+        with open(filepath, encoding="utf-8") as f:
+            data = json.load(f)
+            for domain in data["dial_data"]:
+                # print(domain)
+                for doc_id in data["dial_data"][domain]:
+                    # print(doc_id)
+                    for dialogue in data["dial_data"][domain][doc_id]:
+                        # print(conv_id)
+                        print(dialogue["dial_id"])
+                        for turn in dialogue['turns']:
+                            print(turn["turn_id"])
+                            print(type(turn))
+
+                            for reference in turn['reference']:
+                                print(reference)
+                                print(type(reference))
+
+                                # Features currently used are "context", "question", and "answers".
+                                # Others are extracted here for the ease of future expansions.
+                                yield dialogue["dial_id"], {
+                                        "dial_id":dialogue["dial_id"],
+                                        "domain":domain,
+                                        "doc_id": doc_id,
+                                        "turns" : {
+                                            "turn_id" : turn["turn_id"],
+                                            "role" : turn["role"],
+                                            "da" : turn["da"],
+                                            "reference" : {reference : turn['reference'][reference]},
```

Comment:
I think the dict here must have `keys` and `values` as keys according to the feature types defined above.
Also those two fields must have values that are lists of strings according to the features.

Reply:
Thank @lhoestq! I converted the two fields two lists of strings and added `keys` and `values` as keys however now I'm getting the following error:

```
Traceback (most recent call last):
  File "datasets-cli", line 36, in <module>
    service.run()
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/commands/test.py", line 116, in run
    try_from_hf_gcs=False,
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/builder.py", line 527, in download_and_prepare
    dl_manager=dl_manager, verify_infos=verify_infos, **download_and_prepare_kwargs
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/builder.py", line 604, in _download_and_prepare
    self._prepare_split(split_generator, **prepare_split_kwargs)
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/builder.py", line 905, in _prepare_split
    example = self.info.features.encode_example(record)
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/features.py", line 799, in encode_example
    return encode_nested_example(self, example)
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/features.py", line 710, in encode_nested_example
    (k, encode_nested_example(sub_schema, sub_obj)) for k, (sub_schema, sub_obj) in utils.zip_dict(schema, obj)
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/features.py", line 710, in <genexpr>
    (k, encode_nested_example(sub_schema, sub_obj)) for k, (sub_schema, sub_obj) in utils.zip_dict(schema, obj)
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/features.py", line 728, in encode_nested_example
    list_dict[k] = [encode_nested_example(sub_schema, o) for o in sub_objs]
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/features.py", line 728, in <listcomp>
    list_dict[k] = [encode_nested_example(sub_schema, o) for o in sub_objs]
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/features.py", line 727, in encode_nested_example
    for k, (sub_schema, sub_objs) in utils.zip_dict(schema.feature, obj):
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/utils/py_utils.py", line 111, in zip_dict
    yield key, tuple(d[key] for d in dicts)
  File "/Users/karimfoda/Documents/STUDIES/PYTHON/DATASETS/src/datasets/utils/py_utils.py", line 111, in <genexpr>
    yield key, tuple(d[key] for d in dicts)
KeyError: 'k'
```

I'm certain this has got something to do with the reference dictionary as if I comment that out the tests all pass. I'm not sure what exactly is going wrong with it but the error seems to indicate that it's trying to has the word "keys".

Reply:
I've been investigating this further and I think the issue I'm facing is related to the fact that the json files containing the train & dev data have 4 dictionaries listed within each other (domain - doc_id - turns - reference) and I'm getting errors declaring the lowest level dictionary. (reference). 

The standard `datasets.features.Sequence` doesn't seem to work when it is nested within another `datasets.features.Sequence`. Here's an example of just a 2 turn conversation within the data:

```
{
    "dial_data": {
        "dmv": {
            "Top 5 DMV Mistakes and How to Avoid Them#3_0": [
                {
                    "dial_id": "dea7174409afbfe0af0ace21e7f318ae",
                    "doc_id": "Top 5 DMV Mistakes and How to Avoid Them#3_0",
                    "domain": "dmv",
                    "turns": [
                        {
                            "turn_id": 1,
                            "role": "user",
                            "da": "assert/provide/precondition_pos",
                            "reference": {
                                "23": [
                                    "precondition"
                                ]
                            },
                            "utterance": "My insurance ended so what should i do"
                        },
                        {
                            "turn_id": 2,
                            "role": "agent",
                            "da": "respond/reply/open",
                            "reference": {
                                "24": [
                                    "solution"
                                ],
                                "25": [
                                    "solution"
                                ],
                                "26": [
                                    "solution"
                                ],
                                "23": [
                                    "reference"
                                ]
                            },
                            "utterance": "You will need to get insurance or we will suspend your registration and license"
                        }
                    ]
                }
            ]
        }
    }
}
```